pkg/sentry/devices/memdev/kmsg.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memdev
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/safemem"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserr"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+)
+
+const (
+	BUFFER_ENTRY_MAX = 512
+	BUFFER_MAX       = 1024
+)
+
+// kmsgDevice implements vfs.Device for /dev/kmsg.
+type kmsgDevice struct{}
+
+func (kmsgDevice) Open(ctx context.Context, mnt *vfs.Mount, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	fd := &kmsgFD{newmsg: make(chan int)}
+	if err := fd.vfsfd.Init(fd, opts.Flags, mnt, vfsd, &vfs.FileDescriptionOptions{
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// kmsgFD implements vfs.FileDescriptionImpl for /dev/kmsg.
+type kmsgFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.NoLockFD
+
+	// buffer is a circular buffer that store user's write to kmsg.
+	m            sync.Mutex
+	buffer       [BUFFER_ENTRY_MAX]*bytes.Buffer
+	readIndex    int
+	writeIndex   int
+	bufferLength int
+	newmsg       chan int
+}
+
+// Release implements vfs.FileDescriptionImpl.Release.
+func (fd *kmsgFD) Release() {
+	// noop
+}
+
+// Read implements vfs.FileDescriptionImpl.Read.
+func (fd *kmsgFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+RETRY:
+	fd.m.Lock()
+
+	if fd.bufferLength > len(fd.buffer) {
+		fd.readIndex = fd.writeIndex
+		fd.bufferLength = len(fd.buffer)
+		fd.m.Unlock()
+		return 0, syserr.ErrBrokenPipe.ToError()
+	}
+	if fd.bufferLength == 0 {
+		fd.m.Unlock()
+		if fd.vfsfd.StatusFlags()&^linux.O_NONBLOCK != 0 {
+			return 0, syserror.ErrWouldBlock
+		}
+		<-fd.newmsg
+		goto RETRY
```

Comment:
goto's should generally be the control flow mechanism of last resort. In all of gVisor, there are a tiny handful of goto statements outside of tests: https://cs.opensource.google/search?q=%22goto%20%22%20lang:go&ss=gvisor%2Fgvisor . Gotos can be difficult to reason about, and can lead to subtle scoping bugs.

This could easily be re-written as a for loop. However the whole retry is not necessary, see comment of non-blocking IO.